govcd/user.go

```diff
@@ -0,0 +1,456 @@
+		if role.Name == roleName {
+			return role, nil
+		}
+	}
+
+	return nil, ErrorEntityNotFound
+}
+
+// CreateUser creates an OrgUser from a full configuration structure
+// The timeOut variable is the maximum time we wait for the user to be ready
+// (This operation does not return a task)
+// This function returns as soon as the user has been created, which could be as
+// little as 200ms or as much as Client.MaxRetryTimeout
+// Mandatory fields are: Name, Role, Password.
+// https://code.vmware.com/apis/442/vcloud-director#/doc/doc/operations/POST-CreateUser.html
+func (adminOrg *AdminOrg) CreateUser(userConfiguration *types.User) (*OrgUser, error) {
+	err := validateUserForCreation(userConfiguration)
+	if err != nil {
+		return nil, err
+	}
+
+	userCreateHREF, err := url.ParseRequestURI(adminOrg.AdminOrg.HREF)
+	if err != nil {
+		return nil, fmt.Errorf("error parsing admin org url: %s", err)
+	}
+	userCreateHREF.Path += "/users"
+
+	user := NewUser(adminOrg.client, adminOrg)
+
+	_, err = adminOrg.client.ExecuteRequest(userCreateHREF.String(), http.MethodPost,
+		types.MimeAdminUser, "error creating user: %s", userConfiguration, user.User)
+	if err != nil {
+		return nil, err
+	}
+
+	// If there is a valid task, we try to follow through
+	if user.User.Tasks != nil && len(user.User.Tasks.Task) > 0 {
+		task := NewTask(adminOrg.client)
+		task.Task = user.User.Tasks.Task[0]
+		err = task.WaitTaskCompletion()
+
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	// Attempting to retrieve the user
+	delayPerAttempt := 200 * time.Millisecond
+	maxOperationTimeout := time.Duration(adminOrg.client.MaxRetryTimeout) * time.Second
+
+	// We make sure that the timeout is never less than 2 seconds
+	if maxOperationTimeout < 2*time.Second {
+		maxOperationTimeout = 2 * time.Second
+	}
+
+	// If maxRetryTimeout is set to a higher limit, we lower it to match the
+	// expectations for this operation. If the user is not created within 10 seconds,
+	// there is no need to wait for more. Usually, the operation lasts between 200ms and 900ms
+	if maxOperationTimeout > 10*time.Second {
+		maxOperationTimeout = 10 * time.Second
+	}
+
+	startTime := time.Now()
+	elapsed := time.Since(startTime)
+	var newUser *OrgUser
+	for elapsed < maxOperationTimeout {
+		newUser, err = adminOrg.GetUserByNameOrId(userConfiguration.Name, true)
+		if err == nil {
+			break
+		}
+		time.Sleep(delayPerAttempt)
+		elapsed = time.Since(startTime)
+	}
+
+	elapsed = time.Since(startTime)
+
+	// If the user was not retrieved within the allocated time, we inform the user about the failure
+	// and the time it occurred to get to this point, so that they may try with a longer time
+	if err != nil {
+		return nil, fmt.Errorf("failure to retrieve a new user after %s : %s", elapsed, err)
+	}
+
+	return newUser, nil
+}
+
+// SimpleCreateUser creates an org user from a simplified structure
+func (adminOrg *AdminOrg) SimpleCreateUser(userData OrgUserConfiguration) (*OrgUser, error) {
```

Comment:
I would recommend to inline with CRUD naming. CreateSimpleUser(userData SimpleOrgUserConfiguration)

Reply:
`CreateSimpleUser` sounds to me like we create a user that is less complex than a full one.
In my naming, `Simple` refers to the operation, and that's why it prepends `Create`.

Reply:
I'm also on the boat of those who expect all create operations to begin with `Create...` - it makes it very easy to look at structure of the code.

`CreateSimpleUser` to me sounds fine here, because you are creating a "simple" user, as the param is simplified, but for me `CreateUserSimple` sounds even better, because then it's clear that this is a Create for a User and implies Simple way of doing it. Another synonims for "simple" could be "light" or "plain".

Reply:
Renamed as agreed on chat

Reply:
Reminder to adjust the commit message, because now it (PR description) is still using the old function names.